File: Dynamic Programming/LET_no1312_Minimum Insertion Steps to Make a String Palindrome.py
class Solution:
    def minInsertions(self, s: str) -> int:
        n = len(s)
        dp = [[0 for _ in range(n + 1)] for _ in range(n + 1)]

        for i in range(n):
            for j in range(n):
                if s[i] == s[-1 - j]:
                    dp[i + 1][j + 1] = dp[i][j] + 1
                else:
                    dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j])

        return n - dp[n][n]

File: Dynamic Programming/Hard/LET_no1312_Minimum Insertion Steps to Make a String Palindrome.cpp
class Solution {
public:
    int minInsertions(string s) {
        int n = s.size();
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, 0));

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dp[i + 1][j + 1] = (s[i] == s[n - 1 - j] ? dp[i][j] + 1 : max(dp[i][j + 1], dp[i + 1][j]));
            }
        }

        return n - dp[n][n];
    }
};